lib/sotest.nix

```diff
@@ -3,30 +3,94 @@
 with pkgs.lib;
 
 rec {
-  projectConfigJSON = {
-      boot_prerequisites ? [],
-      extra_dependencies ? [],
-      boot_panic_patterns ? [],
-      boot_items ? []
+  # Create an attribute set of a sotest test run with the format specified in
+  # https://docs.sotest.io/project_config
+  # In order to upload it to a sotest instance, it should be converted to
+  # JSON or YAML, see `asJSONFile` and `asYAMLFile`.
+  projectConfigJSON =
+    { boot_prerequisites ? []
+    , extra_dependencies ? []
+    , boot_panic_patterns ? []
+    , boot_items ? []
     }: builtins.toJSON {
       inherit
         boot_items
         boot_panic_patterns
         boot_prerequisites
-        extra_dependencies;
+        extra_dependencies
+        ;
     };
+
+  # Create a pretty-printed JSON file from the input
   asJSONFile = jsonContent: pkgs.runCommandNoCC "content.json"
     { nativeBuildInputs = [ pkgs.jq ]; }
     "jq '.' ${pkgs.writeText "content.json" jsonContent} > $out";
+
+  # Create a pretty-printed YAML file from the input
   asYAMLFile = jsonContent: pkgs.runCommandNoCC "content.yaml"
     { nativeBuildInputs = [ pkgs.yq ]; }
     "yq -y '.' ${pkgs.writeText "content.yaml" jsonContent} > $out";
-  linuxBootItem = {
-      boot_item_timeout ? 300,
-      name,
-      kernel,
-      kernelParams ? [ "console=@{linux_terminal}" ],
-      initrd
+
+  # Calculate list of direct dependencies of a derivation
+  flatReferences = drv: pkgs.stdenv.mkDerivation {
+    name = "flat-closure-info";
+    __structuredAttrs = true;
+    exportReferencesGraph.closure = [ drv ];
+    preferLocalBuild = true;
+    PATH = "${pkgs.coreutils}/bin:${pkgs.jq}/bin";
+    builder = builtins.toFile "builder" ''
+      . .attrs.sh
+
+      out=''${outputs[out]}
+
+      jq -r '.exportReferencesGraph.closure[0] as $x | .closure[] | select(.path == $x) | .references[]' \
+        < .attrs.json \
+        > $out
+    '';
+  };
+
+  # Generate a ZIP bundle that contains all direct dependencies of a derivation
+  # In case any of these deps have runtime deps themselves, this will *not* work!
+  zipBundleFromTestrunClosure = drv: pkgs.runCommandNoCC "${drv.name}-bundle.zip" {} ''
+    cd /nix/store
+    paths=""
+    for path in $(cat ${flatReferences drv}); do
+      paths="$paths $(basename $path)"
+    done
+    ${pkgs.zip}/bin/zip -r $out $paths
+  '';
+
+  # Pick a subset of paths from within a derivation and create a new derivation
+  # that links to only these paths.
+  # This can be used to reduce the size of a derivation if only parts of it are
```

Comment:
I don't understand how we safe space by linking derivations together. The same set of derivations will be in the closure?

Reply:
I understood it like this:

By default, the zipBundleFromTestrunClosure makes the complete output of a derivation part of the zip file, even if only a small part of it is needed. This function would allow filtering the output paths of a derivation, so it only contains the necessary paths.